src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -94,4 +95,214 @@ public void testOffsetAtRead(String cigar, int expectedReferencePos, int posInRe
             Assert.assertEquals(sam.getReferencePositionAtReadPosition(posInRead), expectedReferencePos);
     }
+
+    @DataProvider(name = "deepCopyTestData")
+    public Object [][] deepCopyTestData() {
+        return new Object[][]{
+                {"3S9M"},
+                {"4M1I6M"},
+        };
+    }
+
+    @Test(dataProvider = "deepCopyTestData")
+    public void testDeepCopyRef(String cigar) {
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, cigar, null, 2);
+        testDeepCopy(sam);
+    }
+
+    @Test(dataProvider = "deepCopyTestData")
+    public void testDeepCopyMutate(String cigar) {
+        final SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, cigar, null, 2);
+        final byte[] initialBaseQualityCopy = Arrays.copyOf(sam.getBaseQualities(), sam.getBaseQualities().length);
+        final int initialStart = sam.getAlignmentStart();
+
+        final SAMRecord deepCopy = testDeepCopy(sam);
+        Assert.assertTrue(Arrays.equals(sam.getBaseQualities(), deepCopy.getBaseQualities()));
+        Assert.assertTrue(sam.getAlignmentStart() == deepCopy.getAlignmentStart());
+
+        // mutate copy and make sure original remains unchanged
+        final byte[] copyBaseQuals = deepCopy.getBaseQualities();
+        for (int i = 0; i < copyBaseQuals.length; i++) {
+            copyBaseQuals[i]++;
+        }
+        deepCopy.setAlignmentStart(initialStart + 1);
+        Assert.assertTrue(Arrays.equals(sam.getBaseQualities(), initialBaseQualityCopy));
+        Assert.assertTrue(sam.getAlignmentStart() == initialStart);
+    }
+
+    @Test(dataProvider = "deepCopyTestData")
+    public void testDeepByteAttributes( final String cigar ) throws Exception {
+        final SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, cigar, null, 2);
+
+        // Note that "samRecord.deepCopy().equals(samRecord)" fails with attributes due to
+        // SAMBinaryTagAndValue.equals using reference equality on attribute values.
+        SAMRecord deepCopy = testDeepCopy(sam);
+        Assert.assertTrue(sam.equals(deepCopy));
+
+        final byte bytes[] = { -2, -1, 0, 1, 2 };
+        sam.setAttribute("BY", bytes);
+        deepCopy = sam.deepCopy();
+
+        // validate reference inequality and content equality
+        final byte samBytes[] = sam.getByteArrayAttribute("BY");
+        final byte copyBytes[] = deepCopy.getByteArrayAttribute("BY");
+        Assert.assertFalse(copyBytes == samBytes);
+        Assert.assertTrue(Arrays.equals(copyBytes, samBytes));
+
+        // validate mutation independence
+        final byte testByte = -1;
+        Assert.assertTrue(samBytes[2] != testByte);  // ensure initial test condition
+        Assert.assertTrue(copyBytes[2] != testByte); // ensure initial test condition
+        samBytes[2] = testByte;                      // mutate original
```

Comment:
Call `sam.setAttribute("BY")` here for good measure (again, just in case `SAMRecord` is ever changed to perform defensive copying of reference return values).